File: .claude/worktrees/agent-ac129a56/FoodDetectionApp/APIService.swift
import Foundation
import UIKit

class APIService {
    static let shared = APIService()

    // MARK: - Configuration
    // In production, set PROXY_BASE_URL to your Firebase Functions URL.
    // The proxy holds the Gemini key server-side so it never ships in the app.
    // For local development, set GEMINI_API_KEY in Secrets.xcconfig (git-ignored).
    private let proxyBaseURL: String? = {
        Bundle.main.infoDictionary?["PROXY_BASE_URL"] as? String
    }()

    private let directApiKey: String? = {
        Bundle.main.infoDictionary?["GEMINI_API_KEY"] as? String
    }()

    private let modelName = "gemini-flash-latest"

    // MARK: - Rate Limiting
    private var lastRequestTime: Date?
    private var dailyRequestCount = 0
    private var dailyCountDate: String = ""

    private func checkRateLimit() throws {
        let now = Date()
        let today = ISO8601DateFormatter().string(from: now).prefix(10)

        if String(today) != dailyCountDate {
            dailyRequestCount = 0
            dailyCountDate = String(today)
        }

        if let last = lastRequestTime, now.timeIntervalSince(last) < 1.0 {
            throw NSError(domain: "RateLimit", code: 429, userInfo: [NSLocalizedDescriptionKey: "Please wait a moment between requests."])
        }

        if dailyRequestCount >= 100 {
            throw NSError(domain: "RateLimit", code: 429, userInfo: [NSLocalizedDescriptionKey: "Daily request limit reached. Try again tomorrow."])
        }

        dailyRequestCount += 1
        lastRequestTime = now
    }

    // MARK: - Retry Logic
    private func performRequest(_ request: URLRequest, retries: Int = 3) async throws -> (Data, URLResponse) {
        var lastError: Error?
        for attempt in 0..<retries {
            do {
                let (data, response) = try await URLSession.shared.data(for: request)
                if let http = response as? HTTPURLResponse, http.statusCode >= 500 && attempt < retries - 1 {
                    try await Task.sleep(nanoseconds: UInt64(pow(2.0, Double(attempt))) * 1_000_000_000)
                    continue
                }
                return (data, response)
            } catch {
                lastError = error
                if attempt < retries - 1 {
                    try await Task.sleep(nanoseconds: UInt64(pow(2.0, Double(attempt))) * 1_000_000_000)
                }
            }
        }
        throw lastError ?? NSError(domain: "APIError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Request failed after retries"])
    }

    // MARK: - Analyze Food
    func analyzeFood(image: UIImage) async throws -> (name: String, info: NutritionInfo) {
        try checkRateLimit()

        guard let base64Image = image.jpegData(compressionQuality: 0.5)?.base64EncodedString() else {
            throw NSError(domain: "ImageError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to encode image"])
        }

        let (data, _) = try await sendAnalyzeFoodRequest(imageBase64: base64Image)
        return try parseNutritionResponse(data: data, fallbackDish: "AI Detected Food")
    }

    // MARK: - Search Food (Text)
    func searchFood(query: String) async throws -> (name: String, info: NutritionInfo) {
        try checkRateLimit()

        let (data, _) = try await sendSearchFoodRequest(query: query)
        return try parseNutritionResponse(data: data, fallbackDish: query)
    }

    // MARK: - AI Coach
    func getCoachAdvice(userStats: UserStats?, logs: [NutritionManager.FoodLog], healthData: String, historyTOON: String, userQuery: String) async throws -> String {
        try checkRateLimit()

        var context = "User Stats: "
        if let stats = userStats {
            context += "Age \(stats.age), \(stats.gender), \(stats.weight)kg, Goal: \(stats.goal). "
        } else {
            context += "Goal: Stay healthy. "
        }

        context += "\nToday's Health: \(healthData)."

        context += "\nToday's Food: "
        if logs.isEmpty {
            context += "Nothing logged yet."
        } else {
            for log in logs.prefix(10) {
                context += "\(log.food) (\(log.calories)kcal), "
            }
        }

        context += "\n\nPast 30 Days History (TOON Format):\n"
        context += historyTOON

        let (data, _) = try await sendCoachRequest(context: context, query: userQuery)
        return parseCoachResponse(data: data)
    }

    // MARK: - Private Request Builders

    private func sendAnalyzeFoodRequest(imageBase64: String) async throws -> (Data, URLResponse) {
        if let proxyURL = proxyBaseURL, !proxyURL.isEmpty {
            // Use backend proxy (production)
            let body: [String: Any] = ["image_base64": imageBase64]
            return try await postJSON(url: "\(proxyURL)/api/v1/analyze-food", body: body)
        } else if let apiKey = directApiKey, !apiKey.isEmpty {
            // Direct Gemini call (development only)
            let promptText = """
            Analyze this food image. Identify the dish.
            Return a JSON object with these exact keys:
            - "Dish": Name of the dish
            - "Calories per 100g": Estimated calories (number as string)
            - "Carbohydrate per 100g": Estimated carbs (number as string)
            - "Protein per 100 gm": Estimated protein (number as string)
            - "Fats per 100 gm": Estimated fats (number as string)
            - "Healthier Recipe": A short advice to make it healthier
            - "Source": "AI Analysis"
            - "micros": A dictionary of key micronutrients (e.g., "Vitamin A": "10%", "Iron": "5mg")
            Return ONLY the JSON.
            """

            let body: [String: Any] = [
                "contents": [["parts": [
                    ["text": promptText],
                    ["inline_data": ["mime_type": "image/jpeg", "data": imageBase64]]
                ]]],
                "generationConfig": ["responseMimeType": "application/json"]
            ]

            return try await postJSON(url: geminiURL(apiKey: apiKey), body: body)
        } else {
            throw NSError(domain: "ConfigError", code: -1, userInfo: [NSLocalizedDescriptionKey: "No API configuration. Set PROXY_BASE_URL or GEMINI_API_KEY."])
        }
    }

    private func sendSearchFoodRequest(query: String) async throws -> (Data, URLResponse) {
        if let proxyURL = proxyBaseURL, !proxyURL.isEmpty {
            let body: [String: Any] = ["query": query]
            return try await postJSON(url: "\(proxyURL)/api/v1/search-food", body: body)
        } else if let apiKey = directApiKey, !apiKey.isEmpty {
            let promptText = """
            Analyze this food query: "\(query)".
            Return a JSON object with these exact keys:
            - "Dish": Name of the dish (be specific based on query)
            - "Calories per 100g": Estimated calories (number as string)
            - "Carbohydrate per 100g": Estimated carbs (number as string)
            - "Protein per 100 gm": Estimated protein (number as string)
            - "Fats per 100 gm": Estimated fats (number as string)
            - "Healthier Recipe": A short advice to make it healthier
            - "Source": "AI Analysis"
            - "micros": A dictionary of key micronutrients
            Return ONLY the JSON.
            """

            let body: [String: Any] = [
                "contents": [["parts": [["text": promptText]]]],
                "generationConfig": ["responseMimeType": "application/json"]
            ]

            return try await postJSON(url: geminiURL(apiKey: apiKey), body: body)
        } else {
            throw NSError(domain: "ConfigError", code: -1, userInfo: [NSLocalizedDescriptionKey: "No API configuration."])
        }
    }

    private func sendCoachRequest(context: String, query: String) async throws -> (Data, URLResponse) {
        if let proxyURL = proxyBaseURL, !proxyURL.isEmpty {
            let body: [String: Any] = ["context": context, "query": query]
            return try await postJSON(url: "\(proxyURL)/api/v1/coach-advice", body: body)
        } else if let apiKey = directApiKey, !apiKey.isEmpty {
            let systemPrompt = """
            You are a friendly and motivating professional health coach.
            You have access to the user's data for the last 30 days.
            Be concise (max 150 words).
            """
            let fullPrompt = "System: \(systemPrompt)\n\nContext:\n\(context)\n\nUser Request: \(query)"

            let body: [String: Any] = [
                "contents": [["parts": [["text": fullPrompt]]]],
                "generationConfig": ["maxOutputTokens": 1000, "temperature": 0.7]
            ]

            return try await postJSON(url: geminiURL(apiKey: apiKey), body: body)
        } else {
            throw NSError(domain: "ConfigError", code: -1, userInfo: [NSLocalizedDescriptionKey: "No API configuration."])
        }
    }

    // MARK: - Helpers

    private func geminiURL(apiKey: String) -> String {
        "https://generativelanguage.googleapis.com/v1beta/models/\(modelName):generateContent?key=\(apiKey)"
    }

    private func postJSON(url: String, body: [String: Any]) async throws -> (Data, URLResponse) {
        guard let endpoint = URL(string: url) else {
            throw NSError(domain: "URLError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await performRequest(request)

        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
            let errorMsg = String(data: data, encoding: .utf8) ?? "Unknown Error"
            print("API Error (\(httpResponse.statusCode)): \(errorMsg)")
            throw NSError(domain: "APIError", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "API Error: \(httpResponse.statusCode)"])
        }

        return (data, response)
    }

    private func parseNutritionResponse(data: Data, fallbackDish: String) throws -> (name: String, info: NutritionInfo) {
        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
           let candidates = json["candidates"] as? [[String: Any]],
           let content = candidates.first?["content"] as? [String: Any],
           let parts = content["parts"] as? [[String: Any]],
           let text = parts.first?["text"] as? String {

            let cleanContent = text
                .replacingOccurrences(of: "```json", with: "")
                .replacingOccurrences(of: "```", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)

            if let jsonData = cleanContent.data(using: .utf8) {
                let info = try JSONDecoder().decode(NutritionInfo.self, from: jsonData)

                var dishName = fallbackDish
                if let dict = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
                   let name = dict["Dish"] as? String {
                    dishName = name
                }

                return (dishName, info)
            }
        }

        throw NSError(domain: "APIError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to parse response"])
    }

    private func parseCoachResponse(data: Data) -> String {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let candidates = json["candidates"] as? [[String: Any]],
              let firstCandidate = candidates.first else {
            return "Coach Error: Unable to parse response."
        }

        var resultText: String?
        if let content = firstCandidate["content"] as? [String: Any],
           let parts = content["parts"] as? [[String: Any]],
           let text = parts.first?["text"] as? String {
            resultText = text
        }

        if let finishReason = firstCandidate["finishReason"] as? String {
            if finishReason == "MAX_TOKENS" {
                return (resultText ?? "") + " [Truncated]"
            }
            if finishReason == "SAFETY" || finishReason == "RECITATION" {
                return "Coach stopped due to Safety filters."
            }
        }

        return resultText ?? "Coach Error: Empty response."
    }
}

File: .claude/worktrees/agent-a1731485/android/app/src/main/java/com/foodsense/android/services/APIService.kt
package com.foodsense.android.services

import android.graphics.Bitmap
import android.util.Base64
import com.foodsense.android.BuildConfig
import com.foodsense.android.data.FoodLog
import com.foodsense.android.data.NutritionInfo
import com.foodsense.android.data.UserStats
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayOutputStream
import java.util.concurrent.TimeUnit

class APIService {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        explicitNulls = false
    }

    // In production, set PROXY_BASE_URL in gradle.properties to your Firebase Functions URL.
    // The proxy holds the Gemini key server-side so it never ships in the app.
    // For local development, set GEMINI_API_KEY in local.properties (git-ignored).
    private val proxyBaseUrl: String? = try { BuildConfig.PROXY_BASE_URL } catch (_: Exception) { null }
    private val directApiKey: String? = try { BuildConfig.GEMINI_API_KEY } catch (_: Exception) { null }
    private val modelName = "gemini-flash-latest"

    // Rate limiting
    private var lastRequestTime = 0L
    private var dailyRequestCount = 0
    private var dailyCountDate = ""

    private fun checkRateLimit() {
        val now = System.currentTimeMillis()
        val today = java.time.LocalDate.now().toString()

        if (today != dailyCountDate) {
            dailyRequestCount = 0
            dailyCountDate = today
        }

        if (now - lastRequestTime < 1000) {
            throw IllegalStateException("Please wait a moment between requests.")
        }

        if (dailyRequestCount >= 100) {
            throw IllegalStateException("Daily request limit reached. Try again tomorrow.")
        }

        dailyRequestCount++
        lastRequestTime = now
    }

    suspend fun analyzeFood(image: Bitmap): Pair<String, NutritionInfo> = withContext(Dispatchers.IO) {
        checkRateLimit()

        val baos = ByteArrayOutputStream()
        image.compress(Bitmap.CompressFormat.JPEG, 70, baos)
        val base64Image = Base64.encodeToString(baos.toByteArray(), Base64.NO_WRAP)

        val raw = if (!proxyBaseUrl.isNullOrBlank()) {
            val body = buildJsonObject {
                put("image_base64", JsonPrimitive(base64Image))
            }
            postJSON("$proxyBaseUrl/api/v1/analyze-food", body)
        } else {
            val promptText = """
                Analyze this food image. Identify the dish.
                Return a JSON object with these exact keys:
                - "Dish": Name of the dish
                - "Calories per 100g": Estimated calories (number as string)
                - "Carbohydrate per 100g": Estimated carbs (number as string)
                - "Protein per 100 gm": Estimated protein (number as string)
                - "Fats per 100 gm": Estimated fats (number as string)
                - "Healthier Recipe": A short advice to make it healthier
                - "Source": "AI Analysis"
                - "micros": A dictionary of key micronutrients
                Return ONLY the JSON.
            """.trimIndent()

            val body = buildJsonObject {
                put("contents", buildJsonArray {
                    add(buildJsonObject {
                        put("parts", buildJsonArray {
                            add(buildJsonObject { put("text", JsonPrimitive(promptText)) })
                            add(buildJsonObject {
                                put("inline_data", buildJsonObject {
                                    put("mime_type", JsonPrimitive("image/jpeg"))
                                    put("data", JsonPrimitive(base64Image))
                                })
                            })
                        })
                    })
                })
                put("generationConfig", buildJsonObject {
                    put("responseMimeType", JsonPrimitive("application/json"))
                })
            }
            postGemini(body)
        }

        parseJsonNutritionResponse(raw)
    }

    suspend fun searchFood(query: String): Pair<String, NutritionInfo> = withContext(Dispatchers.IO) {
        checkRateLimit()

        val raw = if (!proxyBaseUrl.isNullOrBlank()) {
            val body = buildJsonObject {
                put("query", JsonPrimitive(query))
            }
            postJSON("$proxyBaseUrl/api/v1/search-food", body)
        } else {
            val promptText = """
                Analyze this food query: "$query".
                Return a JSON object with these exact keys:
                - "Dish": Name of the dish
                - "Calories per 100g": Estimated calories (number as string)
                - "Carbohydrate per 100g": Estimated carbs (number as string)
                - "Protein per 100 gm": Estimated protein (number as string)
                - "Fats per 100 gm": Estimated fats (number as string)
                - "Healthier Recipe": A short advice to make it healthier
                - "Source": "AI Analysis"
                - "micros": A dictionary of key micronutrients
                Return ONLY the JSON.
            """.trimIndent()

            val body = buildJsonObject {
                put("contents", buildJsonArray {
                    add(buildJsonObject {
                        put("parts", buildJsonArray {
                            add(buildJsonObject { put("text", JsonPrimitive(promptText)) })
                        })
                    })
                })
                put("generationConfig", buildJsonObject {
                    put("responseMimeType", JsonPrimitive("application/json"))
                })
            }
            postGemini(body)
        }

        parseJsonNutritionResponse(raw)
    }

    suspend fun getCoachAdvice(
        userStats: UserStats?,
        logs: List<FoodLog>,
        healthData: String,
        historyToon: String,
        userQuery: String,
    ): String = withContext(Dispatchers.IO) {
        checkRateLimit()

        val context = buildString {
            append("User Stats: ")
            if (userStats != null) {
                append("Age ${userStats.age}, ${userStats.gender}, ${userStats.weight}kg, Goal: ${userStats.goal}. ")
            } else {
                append("Goal: Stay healthy. ")
            }
            append("\nToday's Health: $healthData.")
            append("\nToday's Food: ")
            if (logs.isEmpty()) {
                append("Nothing logged yet.")
            } else {
                logs.take(10).forEach { append("${it.food} (${it.calories}kcal), ") }
            }
            append("\n\nPast 30 Days History (TOON Format):\n")
            append(historyToon)
        }

        val raw = if (!proxyBaseUrl.isNullOrBlank()) {
            val body = buildJsonObject {
                put("context", JsonPrimitive(context))
                put("query", JsonPrimitive(userQuery))
            }
            postJSON("$proxyBaseUrl/api/v1/coach-advice", body)
        } else {
            val systemPrompt = """
                You are a friendly and motivating professional health coach.
                You have access to the user's data for the last 30 days.
                Be concise (max 150 words).
            """.trimIndent()

            val fullPrompt = "System: $systemPrompt\n\nContext:\n$context\n\nUser Request: $userQuery"

            val body = buildJsonObject {
                put("contents", buildJsonArray {
                    add(buildJsonObject {
                        put("parts", buildJsonArray {
                            add(buildJsonObject { put("text", JsonPrimitive(fullPrompt)) })
                        })
                    })
                })
                put("generationConfig", buildJsonObject {
                    put("maxOutputTokens", JsonPrimitive(1000))
                    put("temperature", JsonPrimitive(0.7))
                })
            }
            postGemini(body)
        }

        parseTextResponse(raw)
    }

    // -- Response Parsers (unchanged — both proxy and direct return Gemini's JSON format) --

    private fun parseJsonNutritionResponse(raw: String): Pair<String, NutritionInfo> {
        val payloadText = extractCandidateText(raw)
        val cleanContent = payloadText
            .replace("```json", "")
            .replace("```", "")
            .trim()

        val parsedElement = json.parseToJsonElement(cleanContent)
        val dish = (parsedElement as? JsonObject)?.get("Dish")?.jsonPrimitive?.contentOrNull ?: "AI Detected Food"
        val info = json.decodeFromJsonElement<NutritionInfo>(parsedElement)
        return dish to info
    }

    private fun parseTextResponse(raw: String): String {
        val root = json.parseToJsonElement(raw).jsonObject
        val candidates = root["candidates"] as? JsonArray ?: return "Coach Error: No response candidates."
        val first = candidates.firstOrNull()?.jsonObject ?: return "Coach Error: Empty response."
        val text = first["content"]
            ?.jsonObject
            ?.get("parts")
            ?.jsonArray
            ?.firstOrNull()
            ?.jsonObject
            ?.get("text")
            ?.jsonPrimitive
            ?.contentOrNull

        val finishReason = first["finishReason"]?.jsonPrimitive?.contentOrNull

        if (finishReason == "SAFETY" || finishReason == "RECITATION") {
            return "Coach stopped due to safety filters."
        }
        if (finishReason == "MAX_TOKENS") {
            return "${text.orEmpty()} [Truncated]"
        }
        return text ?: "Coach Error: Response format unrecognized."
    }

    private fun extractCandidateText(raw: String): String {
        val root = json.parseToJsonElement(raw).jsonObject
        val candidates = root["candidates"] as? JsonArray
            ?: throw IllegalStateException("Response missing candidates")

        return candidates.firstOrNull()
            ?.jsonObject
            ?.get("content")
            ?.jsonObject
            ?.get("parts")
            ?.jsonArray
            ?.firstOrNull()
            ?.jsonObject
            ?.get("text")
            ?.jsonPrimitive
            ?.contentOrNull
            ?: throw IllegalStateException("Response missing text")
    }

    // -- HTTP Helpers --

    private fun postJSON(url: String, body: JsonObject): String {
        val requestBody = json.encodeToString(JsonObject.serializer(), body)
            .toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .build()

        return executeWithRetry(request)
    }

    private fun postGemini(body: JsonObject): String {
        val key = directApiKey
        if (key.isNullOrBlank()) {
            throw IllegalStateException("No API configuration. Set PROXY_BASE_URL or GEMINI_API_KEY.")
        }

        val requestBody = json.encodeToString(JsonObject.serializer(), body)
            .toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url("https://generativelanguage.googleapis.com/v1beta/models/$modelName:generateContent?key=$key")
            .post(requestBody)
            .build()

        return executeWithRetry(request)
    }

    private fun executeWithRetry(request: Request, maxRetries: Int = 3): String {
        var lastException: Exception? = null
        repeat(maxRetries) { attempt ->
            try {
                client.newCall(request).execute().use { response ->
                    val raw = response.body?.string().orEmpty()
                    if (response.isSuccessful) return raw
                    if (response.code >= 500 && attempt < maxRetries - 1) {
                        Thread.sleep((1000L * (1 shl attempt)))
                        return@use
                    }
                    throw IllegalStateException("API Error ${response.code}: ${raw.take(200)}")
                }
            } catch (e: Exception) {
                lastException = e
                if (attempt < maxRetries - 1) {
                    Thread.sleep((1000L * (1 shl attempt)))
                }
            }
        }
        throw lastException ?: IllegalStateException("Request failed after $maxRetries retries")
    }
}
